Scala 3.9.0-RC1: trait private val setter generates generic signature that Java subclasses cannot implement

## Compiler version

Scala 3.9.0-RC1

## Minimized code

**MyTrait.scala:**
```scala
package test

trait MyTrait[S]:
  private val myMap: scala.collection.mutable.Map[S, String] = scala.collection.mutable.Map.empty

abstract class MyAbstract[S] extends MyTrait[S]
```

**MyJavaImpl.java:**
```java
package test;

public class MyJavaImpl extends MyAbstract<String> {
}
```

## Steps to reproduce

1. Compile the Scala code with Scala 3.9.0-RC1:
   ```
   scala-cli compile MyTrait.scala --scala 3.9.0-RC1 -d out
   ```

2. Compile the Java class against the output:
   ```
   javac -cp out MyJavaImpl.java
   ```

## Output

```
MyJavaImpl.java:3: error: MyJavaImpl is not abstract and does not override abstract method test$MyTrait$_setter_$test$MyTrait$$myMap_$eq(scala.collection.mutable.Map<java.lang.Object,java.lang.String>) in MyTrait
public class MyJavaImpl extends MyAbstract<String> {
       ^
```

## Expectation

The Java class should compile successfully, as it does with Scala 3.3.8.

## Root cause analysis

The issue is a change in bytecode generation for trait setter methods between Scala 3.3.8 and 3.9.0-RC1.

**Scala 3.3.8** (working) — setter uses **raw type**:
```
// trait MyTrait: abstract setter
public abstract void test$MyTrait$_setter_$test$MyTrait$$myMap_$eq(scala.collection.mutable.Map);
// class MyAbstract: concrete setter
public void test$MyTrait$_setter_$test$MyTrait$$myMap_$eq(scala.collection.mutable.Map);
```
Both signatures match → Java compiler happy.

**Scala 3.9.0-RC1** (broken) — setter uses **generic type**:
```
// trait MyTrait: abstract setter
public abstract void test$MyTrait$_setter_$test$MyTrait$$myMap_$eq(scala.collection.mutable.Map<S, java.lang.String>);
// class MyAbstract: concrete setter
public void test$MyTrait$_setter_$test$MyTrait$$myMap_$eq(scala.collection.mutable.Map<java.lang.Object, java.lang.String>);
```
The abstract setter has `Map<S, String>` but the concrete implementation has `Map<Object, String>` — Java compiler sees these as **different method signatures** and considers the abstract method unimplemented.

## Real-world impact

This breaks **Apache Pekko** (the successor to Akka), where the `FSM` trait has `private val` fields with generic types (e.g., `private val stateTimeouts = mutable.Map[S, Timeout]()`). Java subclasses like `AbstractFSMActorTest.MyFSM extends AbstractFSM<String, String>` fail to compile with Scala 3.9.0-RC1 because the concrete setter in `AbstractFSM` has `Map<Object, ...>` while the abstract setter in `FSM` has `Map<S, ...>`.

This affects multiple Pekko modules: `pekko-actor` (FSM), `pekko-persistence` (PersistentFSMBase), and any user code that extends Scala traits with `private val` fields from Java.

--- a/compiler/src/dotty/tools/backend/jvm/GenericSignatures.scala
+++ b/compiler/src/dotty/tools/backend/jvm/GenericSignatures.scala
@@ -226,9 +226,13 @@ object GenericSignatures {
                 // value classes cannot appear as generic arguments
                 jsig(res, vcBoxing = ValueClassBoxing.Box)
           case _ =>
-            boxedSig(tp.widenDealias.widenNullaryMethod)
+            val typeSym = tp.typeSymbol
+            if typeSym.isTypeParam then
+              typeParamSig(typeSym.name)
+            else
               // `tp` might be a singleton type referring to a getter.
               // Hence the widenNullaryMethod.
+              boxedSig(tp.widenDealias.widenNullaryMethod)
         }
 
       pre.widenDealias match {

--- a/tests/generic-java-signatures/t6344.check
+++ b/tests/generic-java-signatures/t6344.check
@@ -71,6 +71,8 @@ public int C5.f4(java.lang.Integer)
 public int C5.f4(java.lang.Integer)
 public int C5.f5(java.lang.Integer)
 public int C5.f5(java.lang.Integer)
+public int C5.f5(java.lang.Integer,int)
+public int C5.f5(java.lang.Integer,int)
 
 C6
 public java.lang.Object C6.f1(int,java.lang.Object,java.lang.String,java.lang.Object)

--- a/tests/generic-java-signatures/t6344.scala
+++ b/tests/generic-java-signatures/t6344.scala
@@ -28,6 +28,7 @@ class ValA[Q](val value: Q) extends AnyVal {
 }
 class ValB[Q, Q0 <: Q](val value: Q) extends AnyVal {
   def f: Q0 = ???
+  type PQ0 = Q0
 }
 
 class C0 {
@@ -72,6 +73,7 @@ class C5 {
   def f3(x: ValA[Int]) = x.f
   def f4(x: ValB[Int, Int]) = x.f
   def f5(x: ValB[Int, _ <: Int]) = x.f
+  def f5(x: ValB[Int, _ <: Int], y: x.PQ0) = 0
 }
 class C6[A] {
   def f1(x1: Val[A], x2: ValAny[A], x3: ValStr[A], x4: ValA[A]) = x4

--- a/tests/generic-java-signatures/trait-field-generic-signature-parameter.check
+++ b/tests/generic-java-signatures/trait-field-generic-signature-parameter.check
@@ -1,4 +1,4 @@
 public void MyAbstract.MyTrait$_setter_$foo_$eq(scala.collection.mutable.Map)
-public void MyAbstract.MyTrait$_setter_$foo_$eq(scala.collection.mutable.Map<java.lang.Object, java.lang.String>)
+public void MyAbstract.MyTrait$_setter_$foo_$eq(scala.collection.mutable.Map<S, java.lang.String>)
 public scala.collection.mutable.Map MyAbstract.foo()
 public scala.collection.mutable.Map<S, java.lang.String> MyAbstract.foo()

--- a/tests/generic-java-signatures/trait-field-generic-signature-parameter.check
+++ b/tests/generic-java-signatures/trait-field-generic-signature-parameter.check
@@ -1,0 +1,4 @@
+public void MyAbstract.MyTrait$_setter_$foo_$eq(scala.collection.mutable.Map)
+public void MyAbstract.MyTrait$_setter_$foo_$eq(scala.collection.mutable.Map<java.lang.Object, java.lang.String>)
+public scala.collection.mutable.Map MyAbstract.foo()
+public scala.collection.mutable.Map<S, java.lang.String> MyAbstract.foo()

--- a/tests/generic-java-signatures/trait-field-generic-signature-parameter.scala
+++ b/tests/generic-java-signatures/trait-field-generic-signature-parameter.scala
@@ -1,0 +1,11 @@
+trait MyTrait[S]:
+  val foo: scala.collection.mutable.Map[S, String] = scala.collection.mutable.Map.empty
+
+abstract class MyAbstract[S] extends MyTrait[S]
+
+object Test:
+  def main(args: Array[String]): Unit =
+    classOf[MyAbstract[Any]].getMethods.sortBy(_.getName).filter(_.getName.contains("foo")).foreach(m => {
+      println(m)
+      println(m.toGenericString)
+    })

